extension/popup/ui.js

```diff
@@ -0,0 +1,235 @@
+    const anim = lottie.loadAnimation({
+      container, // the dom element that will contain the animation
+      loop,
+      renderer: "svg",
+      autoplay: false,
+      path: `animations/${animationName}.json`, // the path to the animation json
+    });
+    return anim;
+	}
+	
+	function playAnimation(segment, interruptCurrentAnimation, loop) {
+		animation.loop = loop;
+		animation.playSegments(segment, interruptCurrentAnimation);
+	}
+
+  exports.animateByMicVolume = function animateByMicVolume(stream) {
+    animation = loadAnimation("Firefox_Voice_Full", true);
+    const revealAndBase = [
+      animationSegmentTimes.reveal,
+      animationSegmentTimes.base,
+    ];
+    animation.addEventListener("DOMLoaded", function() {
+      animation.playSegments(revealAndBase, true);
+    });
+    getMicVolume(stream, animation);
+  };
+
+  function getMicVolume(stream, animation) {
+    // Inspired by code from https://stackoverflow.com/a/52952907
+    audioContext = new AudioContext();
+    analyser = audioContext.createAnalyser();
+    microphone = audioContext.createMediaStreamSource(stream);
+    javascriptNode = audioContext.createScriptProcessor(2048, 1, 1);
+
+    analyser.smoothingTimeConstant = 0.8;
+    analyser.fftSize = 1024;
+
+    microphone.connect(analyser);
+    analyser.connect(javascriptNode);
+    javascriptNode.connect(audioContext.destination);
+    let updatePoll = 0;
+    // TODO figure out how to stop this update process once we've disabled getUserMedia
+    javascriptNode.onaudioprocess = function() {
+      var array = new Uint8Array(analyser.frequencyBinCount);
+      analyser.getByteFrequencyData(array);
+      let values = 0;
+
+      const length = array.length;
+      for (let i = 0; i < length; i++) {
+        values += array[i];
+      }
+
+      const average = values / length;
+
+      if (updatePoll > 5) {
+        updatePoll = 0;
+        // console.log("this is how frequently i'm updating");
+        setAnimationForVolume(average, animation);
+      } else {
+        updatePoll++;
+      }
+    };
+  }
+
+  function setAnimationForVolume(avgVolume, animation) {
+    console.log(`updating: volume at ${avgVolume}`);
+    animation.onLoopComplete = function() {
+      // code here
+      if (avgVolume < 5) {
+				playAnimation(animationSegmentTimes.base, true, true)
+      } else if (avgVolume < 10) {
+				playAnimation(animationSegmentTimes.low, true, true)
+      } else if (avgVolume < 20) {
+				playAnimation(animationSegmentTimes.medium, true, true)
+      } else {
+				playAnimation(animationSegmentTimes.high, true, true)
+      }
+    };
+  }
+
+  // Event handler for when we detect the user has started typing
+  function detectText(e) {
+    if (!textInputDetected) {
+      exports.setState("typing"); // TODO: is this the right place to set the state? or should that all be handled by popup.js
+      textInputDetected = true;
+      // TODO: SEND MESSAGE TO STOP MIC INPUT
```

Comment:
I guess this could call `exports.onStartTextInput()`, and then popup.js will override that (which is a little ugly, but find for now)